src/dag/write.ts

```diff
@@ -85,64 +87,67 @@ export class Write extends Read {
   }
 
   async commit(): Promise<void> {
-    await this.collectGarbage();
+    await this._collectGarbage();
     await this._tx.commit();
   }
 
-  async collectGarbage(): Promise<void> {
+  private async _collectGarbage(): Promise<void> {
     // We increment all the ref counts before we do all the decrements. This
     // is so that we do not remove an item that goes from 1 -> 0 -> 1
-    const newHeads: (Hash | undefined)[] = [];
-    const oldHeads: (Hash | undefined)[] = [];
+    const newHeads: Hash[] = [];
+    const oldHeads: Hash[] = [];
     for (const changedHead of this._changedHeads.values()) {
-      oldHeads.push(changedHead.old);
-      newHeads.push(changedHead.new);
+      changedHead.old && oldHeads.push(changedHead.old);
+      changedHead.new && newHeads.push(changedHead.new);
     }
 
     for (const n of newHeads) {
-      if (n !== undefined) {
-        await this.changeRefCount(n, 1);
-      }
+      await this._changeRefCount(n, 1);
     }
 
     for (const o of oldHeads) {
-      if (o !== undefined) {
-        await this.changeRefCount(o, -1);
-      }
+      await this._changeRefCount(o, -1);
     }
 
     // Now we go through the mutated chunks to see if any of them are still orphaned.
     const ps = [];
     for (const hash of this._newChunks) {
-      const count = await this.getRefCount(hash);
+      const count = await this._getRefCount(hash);
       if (count === 0) {
-        ps.push(this.removeAllRelatedKeys(hash, false));
+        ps.push(this._removeAllRelatedKeys(hash, false));
       }
     }
     await Promise.all(ps);
   }
 
-  async changeRefCount(hash: Hash, delta: number): Promise<void> {
-    const oldCount = await this.getRefCount(hash);
-    const newCount = oldCount + delta;
-
-    if ((oldCount === 0 && delta === 1) || (oldCount === 1 && delta === -1)) {
-      const meta = await this._tx.get(chunkMetaKey(hash));
-      if (meta !== undefined) {
-        assertMeta(meta);
-        const ps = meta.map(ref => this.changeRefCount(ref, delta));
-        await Promise.all(ps);
-      }
+  private async _changeRefCount(hash: Hash, delta: number): Promise<void> {
+    let lock = this._changeRefCountLocks.get(hash);
```

Comment:
It might work to have a sync cache for the ref count instead. Trying that

Reply:
Interesting and subtle bug. Yeah my mind went to same place as yours -- can we sort of parallelize getting all the ref counts that we will need in one pass, change them as needed, then flush them in a different pass? Would tha t simplify things?

Reply:
That seems to work